lib/gpt.js

```
@@ -518,6 +526,56 @@ GPT.prototype = {
 
     return buffer
 
+  },
+
+  /**
+   * Relocate the backup GPT to another LBA
+   *
+   * According to the UEFI spec, the backup GPT should always be located
+   * at the end of the disk. For us, it means that we only support moving
+   * it after its current location.
+   *
+   * @param {Number} [newLBA] - this is the LBA of the backup GPT header
+   */
+  relocate( newLBA ) {
+    // Check whether this is a primary or a backup GPT
+    var isPrimary = this.backupLBA == 0 || this.currentLBA < this.backupLBA
+
+    // Set the new LBA of the backup GPT
+    if( isPrimary ) {
+
+      if( newLBA <= this.backupLBA ) {
+        return
```

I wonder if this should raise an error, rather than just silently returning? :shrug:

---

Throwing an error would be a bit much here I think; but returning the offset in bytes it was relocated to would be good, so that one can check and assert on that return value @etcher123

---

Not sure this behavior is as helpful as it might seem at first glance – i.e. if you want to restore a primary GPT from a secondary, you'd need to relocate the secondary forwards, to the start of the device, and this would prevent doing that easily.

---

Feels to me like a hypothetical `createPrimaryFromSecondary` is actually something different to `relocateSecondary` ? :shrug: 

I agree that an error might be a bit much - I just meant that the caller should have some way of knowing whether this function succeeded or not, rather than "silently" doing nothing. And returning the new offset would do that :+1:

---

Just as a reminder, the original purpose of this PR was to allow relocating the secondary GPT to the end of the disk, a specific use-case tailored for installers (like Etcher) that install an image of fixed size on a disk of unknown size (equal or bigger than the image). So the `relocate()` function proposed here did just that originally.

@jhermsmeier I understand you want to make it more generic. But how much?

We can make it "relocate the secondary GPT anywhere with no check", so basically I just remove the checks `newLBA <= this.backupLBA` and `newLBA <= this.currentLBA` and that's it. I'm fine with that, but is there any use-case for that? (Note that, even without obvious use-case, I'm still fine removing the checks, and assume user takes care of not shooting himself in the foot).

I'm not sure to understand the other use-case that you mention though:

> f you want to restore a primary GPT from a secondary, you'd need to relocate the secondary forwards, to the start of the device, and this would prevent doing that easily.

If you'd want to do that, the way I understand it, it would requires two steps:
- `createPrimaryFromSecondary` as @lurch suggests (or maybe a better name would be `writePrimaryFromBackup` to match what already exists in the code). That basically would give you a primary GPT.
- Then just write it at the beginning of the disk? And that's all? There would be nothing else to do.

---

> Throwing an error would be a bit much here I think; but returning the offset in bytes it was relocated to would be good, so that one can check and assert on that return value @etcher123

@jhermsmeier I just did that, see https://github.com/jhermsmeier/node-gpt/compare/c04d81a9ec338edf3460cd4192de52ed4747a73d..9f9d0cb9444dd4194f0d2ee0ced83a50b1b78563